File: Editor/Tools/SetRectTransformTool.cs
using System;
using System.Collections.Generic;
using McpUnity.Unity;
using Newtonsoft.Json.Linq;
using UnityEditor;
using UnityEngine;

namespace McpUnity.Tools
{
    /// <summary>
    /// Tool for setting RectTransform layout properties on UI GameObjects.
    /// Supports named presets and raw property overrides.
    /// </summary>
    public class SetRectTransformTool : McpToolBase
    {
        private struct RectTransformPreset
        {
            public Vector2 AnchorMin;
            public Vector2 AnchorMax;
            public Vector2 Pivot;
            public Vector2 AnchoredPosition;
            public Vector2 SizeDelta;

            public RectTransformPreset(Vector2 anchorMin, Vector2 anchorMax, Vector2 pivot, Vector2 anchoredPosition, Vector2 sizeDelta)
            {
                AnchorMin = anchorMin;
                AnchorMax = anchorMax;
                Pivot = pivot;
                AnchoredPosition = anchoredPosition;
                SizeDelta = sizeDelta;
            }
        }

        private static readonly Dictionary<string, RectTransformPreset> Presets = new Dictionary<string, RectTransformPreset>
        {
            ["stretch"] = new RectTransformPreset(new Vector2(0f, 0f), new Vector2(1f, 1f), new Vector2(0.5f, 0.5f), new Vector2(0f, 0f), new Vector2(0f, 0f)),
            ["center"] = new RectTransformPreset(new Vector2(0.5f, 0.5f), new Vector2(0.5f, 0.5f), new Vector2(0.5f, 0.5f), new Vector2(0f, 0f), new Vector2(100f, 100f)),
            ["top-left"] = new RectTransformPreset(new Vector2(0f, 1f), new Vector2(0f, 1f), new Vector2(0f, 1f), new Vector2(0f, 0f), new Vector2(100f, 100f)),
            ["top-center"] = new RectTransformPreset(new Vector2(0.5f, 1f), new Vector2(0.5f, 1f), new Vector2(0.5f, 1f), new Vector2(0f, 0f), new Vector2(100f, 100f)),
            ["top-right"] = new RectTransformPreset(new Vector2(1f, 1f), new Vector2(1f, 1f), new Vector2(1f, 1f), new Vector2(0f, 0f), new Vector2(100f, 100f)),
            ["middle-left"] = new RectTransformPreset(new Vector2(0f, 0.5f), new Vector2(0f, 0.5f), new Vector2(0f, 0.5f), new Vector2(0f, 0f), new Vector2(100f, 100f)),
            ["middle-right"] = new RectTransformPreset(new Vector2(1f, 0.5f), new Vector2(1f, 0.5f), new Vector2(1f, 0.5f), new Vector2(0f, 0f), new Vector2(100f, 100f)),
            ["bottom-left"] = new RectTransformPreset(new Vector2(0f, 0f), new Vector2(0f, 0f), new Vector2(0f, 0f), new Vector2(0f, 0f), new Vector2(100f, 100f)),
            ["bottom-center"] = new RectTransformPreset(new Vector2(0.5f, 0f), new Vector2(0.5f, 0f), new Vector2(0.5f, 0f), new Vector2(0f, 0f), new Vector2(100f, 100f)),
            ["bottom-right"] = new RectTransformPreset(new Vector2(1f, 0f), new Vector2(1f, 0f), new Vector2(1f, 0f), new Vector2(0f, 0f), new Vector2(100f, 100f)),
            ["stretch-horizontal"] = new RectTransformPreset(new Vector2(0f, 0.5f), new Vector2(1f, 0.5f), new Vector2(0.5f, 0.5f), new Vector2(0f, 0f), new Vector2(0f, 100f)),
            ["stretch-vertical"] = new RectTransformPreset(new Vector2(0.5f, 0f), new Vector2(0.5f, 1f), new Vector2(0.5f, 0.5f), new Vector2(0f, 0f), new Vector2(100f, 0f))
        };

        public SetRectTransformTool()
        {
            Name = "set_rect_transform";
            Description = "Sets RectTransform layout properties on a UI GameObject using presets and/or raw property overrides.";
            IsAsync = false;
        }

        public override JObject Execute(JObject parameters)
        {
            var findResult = TransformToolUtils.FindGameObject(parameters);
            if (findResult.Error != null)
            {
                return findResult.Error;
            }

            GameObject gameObject = findResult.GameObject;
            RectTransform rectTransform = gameObject.GetComponent<RectTransform>();
            if (rectTransform == null)
            {
                return McpUnitySocketHandler.CreateErrorResponse(
                    $"GameObject '{gameObject.name}' does not have a RectTransform component",
                    "validation_error"
                );
            }

            string presetName = parameters["preset"]?.ToObject<string>();
            JObject anchoredPositionObj = parameters["anchoredPosition"] as JObject;
            JObject sizeDeltaObj = parameters["sizeDelta"] as JObject;
            JObject anchorMinObj = parameters["anchorMin"] as JObject;
            JObject anchorMaxObj = parameters["anchorMax"] as JObject;
            JObject pivotObj = parameters["pivot"] as JObject;
            JObject offsetMinObj = parameters["offsetMin"] as JObject;
            JObject offsetMaxObj = parameters["offsetMax"] as JObject;

            bool hasPreset = !string.IsNullOrWhiteSpace(presetName);
            bool hasRawOverrides = anchoredPositionObj != null
                                   || sizeDeltaObj != null
                                   || anchorMinObj != null
                                   || anchorMaxObj != null
                                   || pivotObj != null
                                   || offsetMinObj != null
                                   || offsetMaxObj != null;

            if (!hasPreset && !hasRawOverrides)
            {
                return McpUnitySocketHandler.CreateErrorResponse(
                    "At least one of 'preset', 'anchoredPosition', 'sizeDelta', 'anchorMin', 'anchorMax', 'pivot', 'offsetMin', or 'offsetMax' must be provided",
                    "validation_error"
                );
            }

            // Validate preset name before recording undo (avoid undo noise on validation failures)
            RectTransformPreset? validatedPreset = null;
            if (hasPreset)
            {
                string normalizedPreset = NormalizePresetName(presetName);
                if (!Presets.TryGetValue(normalizedPreset, out RectTransformPreset preset))
                {
                    return McpUnitySocketHandler.CreateErrorResponse(
                        $"Invalid preset '{presetName}'. Valid presets: {string.Join(", ", Presets.Keys)}",
                        "validation_error"
                    );
                }
                validatedPreset = preset;
            }

            Undo.RecordObject(rectTransform, "Set RectTransform");

            if (validatedPreset.HasValue)
            {
                var preset = validatedPreset.Value;
                rectTransform.anchorMin = preset.AnchorMin;
                rectTransform.anchorMax = preset.AnchorMax;
                rectTransform.pivot = preset.Pivot;
                rectTransform.anchoredPosition = preset.AnchoredPosition;
                rectTransform.sizeDelta = preset.SizeDelta;
            }

            if (anchoredPositionObj != null)
            {
                rectTransform.anchoredPosition = ApplyVector2Override(rectTransform.anchoredPosition, anchoredPositionObj);
            }

            if (sizeDeltaObj != null)
            {
                rectTransform.sizeDelta = ApplyVector2Override(rectTransform.sizeDelta, sizeDeltaObj);
            }

            if (anchorMinObj != null)
            {
                rectTransform.anchorMin = ApplyVector2Override(rectTransform.anchorMin, anchorMinObj);
            }

            if (anchorMaxObj != null)
            {
                rectTransform.anchorMax = ApplyVector2Override(rectTransform.anchorMax, anchorMaxObj);
            }

            if (pivotObj != null)
            {
                rectTransform.pivot = ApplyVector2Override(rectTransform.pivot, pivotObj);
            }

            if (offsetMinObj != null)
            {
                rectTransform.offsetMin = ApplyVector2Override(rectTransform.offsetMin, offsetMinObj);
            }

            if (offsetMaxObj != null)
            {
                rectTransform.offsetMax = ApplyVector2Override(rectTransform.offsetMax, offsetMaxObj);
            }

            EditorUtility.SetDirty(gameObject);

            return new JObject
            {
                ["success"] = true,
                ["type"] = "text",
                ["message"] = $"RectTransform updated successfully for GameObject '{gameObject.name}'.",
                ["data"] = new JObject
                {
                    ["instanceId"] = gameObject.GetInstanceID(),
                    ["name"] = gameObject.name,
                    ["path"] = TransformToolUtils.GetGameObjectPath(gameObject),
                    ["anchorMin"] = Vector2ToJObject(rectTransform.anchorMin),
                    ["anchorMax"] = Vector2ToJObject(rectTransform.anchorMax),
                    ["pivot"] = Vector2ToJObject(rectTransform.pivot),
                    ["anchoredPosition"] = Vector2ToJObject(rectTransform.anchoredPosition),
                    ["sizeDelta"] = Vector2ToJObject(rectTransform.sizeDelta),
                    ["offsetMin"] = Vector2ToJObject(rectTransform.offsetMin),
                    ["offsetMax"] = Vector2ToJObject(rectTransform.offsetMax)
                }
            };
        }

        private static string NormalizePresetName(string presetName)
        {
            string[] parts = presetName.Trim().ToLowerInvariant().Split(new[] { '-', '_', ' ' }, StringSplitOptions.RemoveEmptyEntries);
            return string.Join("-", parts);
        }

        private static Vector2 ApplyVector2Override(Vector2 current, JObject value)
        {
            return new Vector2(
                value["x"]?.ToObject<float>() ?? current.x,
                value["y"]?.ToObject<float>() ?? current.y
            );
        }

        private static JObject Vector2ToJObject(Vector2 vector)
        {
            return new JObject
            {
                ["x"] = vector.x,
                ["y"] = vector.y
            };
        }
    }
}

File: Server~/src/tools/setRectTransformTool.ts
import * as z from 'zod';
import { Logger } from '../utils/logger.js';
import { McpUnity } from '../unity/mcpUnity.js';
import { McpServer } from '@modelcontextprotocol/sdk/server/mcp.js';
import { McpUnityError, ErrorType } from '../utils/errors.js';
import { CallToolResult } from '@modelcontextprotocol/sdk/types.js';

const toolName = 'set_rect_transform';
const toolDescription =
  'Sets RectTransform layout properties on a UI GameObject. Supports presets (stretch, center, top-left, top-center, top-right, middle-left, middle-right, bottom-left, bottom-center, bottom-right, stretch-horizontal, stretch-vertical) and optional raw overrides (anchoredPosition, sizeDelta, anchorMin, anchorMax, pivot, offsetMin, offsetMax). Preset is applied first, then raw overrides.';

const vector2Schema = z
  .object({
    x: z.number().optional(),
    y: z.number().optional()
  })
  .describe('Vector2 with optional x/y components');

const presetSchema = z.enum([
  'stretch',
  'center',
  'top-left',
  'top-center',
  'top-right',
  'middle-left',
  'middle-right',
  'bottom-left',
  'bottom-center',
  'bottom-right',
  'stretch-horizontal',
  'stretch-vertical'
]);

const paramsSchema = z.object({
  instanceId: z.number().optional().describe('The instance ID of the target GameObject'),
  objectPath: z.string().optional().describe('The hierarchy path of the target GameObject (alternative to instanceId)'),
  preset: presetSchema.optional().describe('Optional layout preset to apply first'),
  anchoredPosition: vector2Schema.optional().describe('Optional anchoredPosition override'),
  sizeDelta: vector2Schema.optional().describe('Optional sizeDelta override'),
  anchorMin: vector2Schema.optional().describe('Optional anchorMin override'),
  anchorMax: vector2Schema.optional().describe('Optional anchorMax override'),
  pivot: vector2Schema.optional().describe('Optional pivot override'),
  offsetMin: vector2Schema.optional().describe('Optional offsetMin override'),
  offsetMax: vector2Schema.optional().describe('Optional offsetMax override')
});

export function registerSetRectTransformTool(server: McpServer, mcpUnity: McpUnity, logger: Logger) {
  logger.info(`Registering tool: ${toolName}`);

  server.tool(toolName, toolDescription, paramsSchema.shape, async (params: any) => {
    try {
      logger.info(`Executing tool: ${toolName}`, params);
      const result = await toolHandler(mcpUnity, params);
      logger.info(`Tool execution successful: ${toolName}`);
      return result;
    } catch (error) {
      logger.error(`Tool execution failed: ${toolName}`, error);
      throw error;
    }
  });
}

async function toolHandler(mcpUnity: McpUnity, params: any): Promise<CallToolResult> {
  validateGameObjectIdentifier(params);

  // Validate at least one layout input is provided
  const hasLayoutInput = params.preset !== undefined ||
    params.anchoredPosition !== undefined || params.sizeDelta !== undefined ||
    params.anchorMin !== undefined || params.anchorMax !== undefined ||
    params.pivot !== undefined || params.offsetMin !== undefined ||
    params.offsetMax !== undefined;

  if (!hasLayoutInput) {
    throw new McpUnityError(
      ErrorType.VALIDATION,
      "At least one of 'preset', 'anchoredPosition', 'sizeDelta', 'anchorMin', 'anchorMax', 'pivot', 'offsetMin', or 'offsetMax' must be provided"
    );
  }

  const response = await mcpUnity.sendRequest({
    method: toolName,
    params: {
      instanceId: params.instanceId,
      objectPath: params.objectPath,
      preset: params.preset,
      anchoredPosition: params.anchoredPosition,
      sizeDelta: params.sizeDelta,
      anchorMin: params.anchorMin,
      anchorMax: params.anchorMax,
      pivot: params.pivot,
      offsetMin: params.offsetMin,
      offsetMax: params.offsetMax
    }
  });

  if (!response.success) {
    throw new McpUnityError(
      ErrorType.TOOL_EXECUTION,
      response.message || 'Failed to set RectTransform'
    );
  }

  // Include structured RectTransform state in response for programmatic use
  let text = response.message || 'RectTransform updated';
  if (response.data?.rectTransform) {
    const rt = response.data.rectTransform;
    text += `\n\nCurrent state:`;
    text += `\n  anchoredPosition: (${rt.anchoredPosition.x}, ${rt.anchoredPosition.y})`;
    text += `\n  sizeDelta: (${rt.sizeDelta.x}, ${rt.sizeDelta.y})`;
    text += `\n  anchorMin: (${rt.anchorMin.x}, ${rt.anchorMin.y})`;
    text += `\n  anchorMax: (${rt.anchorMax.x}, ${rt.anchorMax.y})`;
    text += `\n  pivot: (${rt.pivot.x}, ${rt.pivot.y})`;
  }

  return {
    content: [
      {
        type: response.type,
        text
      }
    ]
  };
}

function validateGameObjectIdentifier(params: { instanceId?: number; objectPath?: string }) {
  if (
    (params.instanceId === undefined || params.instanceId === null) &&
    (!params.objectPath || params.objectPath.trim() === '')
  ) {
    throw new McpUnityError(ErrorType.VALIDATION, "Either 'instanceId' or 'objectPath' must be provided");
  }
}
